packages/react-instantsearch-core/src/core/createInstantSearchManager.js

```diff
@@ -296,14 +321,43 @@ export default function createInstantSearchManager({
       return;
     }
 
-    if (!client._useCache || typeof client.addAlgoliaAgent !== 'function') {
-      // This condition avoids hydrating a `searchClient` different from the
-      // Algolia one. We also avoid to hydrate the client when the cache is
-      // disabled. The implementation is brittle but we don't have a proper way
-      // to detect the Algolia client at the moment.
+    // Disable cache hydration on:
+    // - Algoliasearch API Client < v4 with cache disabled
+    // - Third party clients (detected by the `addAlgoliaAgent` function missing)
+    if (
+      !client.transporter &&
+      (!client._useCache || typeof client.addAlgoliaAgent !== 'function')
+    ) {
       return;
     }
 
+    // Algoliasearch API Client >= v4
+    // To hydrate the client we need to populate the cache with the data from
+    // the server (done in `hydrateSearchClientWithMultiIndexRequest` or
+    // `hydrateSearchClientWithSingleIndexRequest`). But since there is no way
+    // for us to compute the key the same way as `algoliasearch-client` we need
+    // to populate it on a custom key and override the `search` method to
+    // search on it first.
+    if (client.transporter) {
+      const baseMethod = client.search;
+      client.search = (requests, ...methodArgs) => {
+        const requestsWithSeriazizedParams = requests.map(request => ({
+          ...request,
+          params: serializeQueryParameters(request.params),
+        }));
+
+        return client.transporter.responsesCache.get(
```

Comment:
responsesCache can be undefined

Reply:
@Haroenv We have a `if (client.transporter)` to ensure we on the `v4` where the responsesCache can't be undefined.

Reply:
unless someone has a transporter which is an empty object as their custom client. It's not likely, but technically possible. Up to us to then decide if we can assume people won't do that

Reply:
I would say that scenario will never happen. What are your thoughts @yannickcr?

Reply:
I agree that it is unlikely to happen. But I can rework the conditions here to make it clearer.

What is missing here is a proper way to know if we are using the `official` client and to get it's version. For now we detect if `client.addAlgoliaAgent` exists or not (does not exists -> `custom`) and we planned to add a `version` property to the client `>= v4` (does not exists -> `<= v4`) to replace the check on `client.transporter`. We could put a more obscure property to avoid any unintentional clash like `__ALGOLIA_CLIENT_VERSION` (a Symbol() would work perfectly for this but IE11...)

Reply:
We could have a `name` property on the client with the value `algoliasearch`.

`if (client.name === 'algoliasearch' && client.version.charAt(0) === 4)`

Reply:
That's also a possibility

Reply:
Let's improve this later, I think for now it's fine to keep this.